### ISSUE TITLE
[ds.channum for ds in data] not in ascending order

### ISSUE THREAD

**Issue by joefowler** [usnistgov/mass#95]
**[Original report](https://bitbucket.org/joe_fowler/mass/issues/95) by Galen O'Neil (Bitbucket: [oneilg](https://bitbucket.org/oneilg), GitHub: [oneilg](https://github.com/oneilg)).**

----------------------------------------

The iteration across datasets is isn't in ascending order anymore as of 0.5.2 at least. It was in ascending order before. Sang Jun thinks 0.5.1 was in ascending order.

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Can you provide a test case? By *ascending order*, do you mean ascending channel number?

Does this only happen when you construct a `TESGroup` using the "glob" mechanism? That is (and here I'm only speculating), does the iterator go in lexical order by filename:

```
data = mass.TESGroup("/data/newdata/files_chan*.ljh", "/data/newdata/noise_chan*.noi")
print [ds.channum for ds in data]
```

Or did this bug arise when you looked at a data set in HDF5-only format?

**Comment by joefowler** [comment]
**Original comment by Galen O'Neil (Bitbucket: [oneilg](https://bitbucket.org/oneilg), GitHub: [oneilg](https://github.com/oneilg)).**

----------------------------------------

I'll try to be more concrete, now that you mention it, I suspect it is likely to be an HDF5 only thing.

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Yeah, looking at the HDF5-only constructor, I see that it builds the important structure, `self.datasets` in the same order as the top-level HDF5 file-group values are yielded by the `keys()` method (which is, of course, an unspecified order).

Omitting some code from `channel_group_hdf5_only.py`:

```
class TESGroupHDF5(channel_group.TESGroup):
    def __init__(self, h5master_fname):
        self.hdf5_file = h5py.File(h5master_fname,"a")

        dset_list = []
        for key in self.hdf5_file.keys():
            if not key.startswith("chan"):
                continue
            grp = self.hdf5_file[key]
            # Omitting something verbose
            dset_list.append(channel.MicrocalDataSet(pulserec_dict, tes_group=self, hdf5_group=grp))
```

I'll fix this problem.

**Comment by joefowler** [comment]
**Original comment by Galen O'Neil (Bitbucket: [oneilg](https://bitbucket.org/oneilg), GitHub: [oneilg](https://github.com/oneilg)).**

----------------------------------------

I found the same thing. f26e090 should fix it.

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Fixes dataset ordering in a TESGroupHDF5

Fixes #95. Galen’s fix didn’t actually work.

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

p.s. I added a test for this, by hard-linking to the regression-test LJH file multiple times (in a temporary directory). We can verify that the sorting works, both on normal `TESGroup` objects and on the problematic one, the `TESGroupHDF5`.
